File: 그래프_dfs/leetcode2359(Medium)-Find Closest Node to Given Two Nodes.py
# 2168ms(28.16%), 151.4MB(13.79%)
class Solution:
    def closestMeetingNode(self, edges: List[int], node1: int, node2: int) -> int:
        N = len(edges)
        graph = {}
        for v, w in enumerate(edges):
            graph[v] = w

        def dfs(v, visited, dist, depth):
            visited[v] = True
            dist[v] = depth
            w = graph[v]
            if not visited[w] and w != -1:
                dfs(w, visited, dist, depth + 1)

        visited1 = [False] * N
        dist1 = [0] * N
        dfs(node1, visited1, dist1, 0)

        visited2 = [False] * N
        dist2 = [0] * N
        dfs(node2, visited2, dist2, 0)

        min_dist = float('inf')
        result = -1
        for v in range(N):
            current_max = max(dist1[v], dist2[v])
            if visited1[v] and visited2[v] and min_dist > current_max:
                min_dist = current_max
                result = v

        return result

File: 그래프_dfs/leetcode2359(Medium)-Find Closest Node to Given Two Nodes.js
// 177ms(85.71%), 81.2MB(42.86%)
/**
 * @param {number[]} edges
 * @param {number} node1
 * @param {number} node2
 * @return {number}
 */
var closestMeetingNode = function (edges, node1, node2) {
  const N = edges.length
  const graph = edges.reduce((obj, w, v) => {
    obj[v] = w
    return obj
  }, {})

  const dfs = (v, visited, dist, depth) => {
    visited[v] = true
    dist[v] = depth
    const w = graph[v]
    if (!visited[w] && w !== -1) {
      dfs(w, visited, dist, depth + 1)
    }
  }

  const visited1 = new Array(N).fill(false)
  const dist1 = new Array(N).fill(0)

  const visited2 = new Array(N).fill(false)
  const dist2 = new Array(N).fill(0)

  dfs(node1, visited1, dist1, 0)
  dfs(node2, visited2, dist2, 0)

  let minDist = Infinity
  let result = -1
  for (let v = 0; v < N; v++) {
    const currentMax = Math.max(dist1[v], dist2[v])
    if (visited1[v] && visited2[v] && minDist > currentMax) {
      minDist = currentMax
      result = v
    }
  }

  return result
}
